### ISSUE TITLE
Feature request: Synchronising with git

### ISSUE THREAD

**Issue by AaronTraas** [laurent22/joplin#3670]
I like Joplin, and am looking for a OneNote replacement. I don't want to have to setup WebDAV on a server of mine, nor do I like the idea of using OneDrive or DropBox. 

Ideally would be git as a sync backend. I'd love to see diffs of all my changes and such as well. 

If this is in-line with the goals of the project, I am willing to contribute code. I just don't want to waste my time if the maintainers here would reject this idea because it doesn't line up with the goals of the project.

**Comment by CalebJohn** [comment]
This is a duplicate of #753. It seems that the original was labeled as an enhancement, which means it's probably not entirely out of line. 

I would guess that the viability of this pull request probably hinges on the complexity of adding it. If it's fairly simple I imagine that it will get picked up for sure, but if it requires sweeping changes to the sync logic then it's probably DoA. 


@laurent22 would you accept a pull request for this?

**Comment by laurent22** [comment]
In theory I'll accept it but in practice I'm not sure there's any easy way to add this while supporting both mobile and desktop clients, but if @AaronTraas has some ideas on how to implement this I'd be glad to hear them.

**Comment by asprionj** [comment]
Hi, first of all: very nice tool, thanks! I'm also about to use a self-hosted git-server (Gitea) for sync between multiple remote users (by having a local file-system sync from Joplin to a folder which is a local git image of the remote repo on the git server). We will see how this works out; merging conflicts would be done on the git / git-server level and should not be too difficult...? AFAIK for changes on an existing note only the file of the note itself changes?

Of course, git integration would be awesome. Indicating, by colors etc., which notes have changed, are new, etc. Diffing, staging, committing would not even be necessary from within Joplin (but very welcome as well, of course). But even getting the info on the status of each file would require Joplin to invoke system-installed git, would it? Or could [isomorphic-git](https://github.com/isomorphic-git/isomorphic-git) help? Don't know Joplin's general architecture at all, but if it can sync with a local folder, it should also be able to run some git commands on that folder...?

**Comment by CalebJohn** [comment]
> AFAIK for changes on an existing note only the file of the note itself changes?

 the sync target notes contain metadata (date etc.)  So that will potentially be updated when you edit a note.

I'm not sure exactly what you're asking in the second part. But I can give you some info that might help. Joplin implements it's own sync and revision system, this works by assuming a stable folder either on a local system or remotely accessible via API (e.g.webdav). This does not immediately conflict with syncing to a got client because that basic assumption is still met. However the current Joplin sync code does not exactly resemble the git architecture and to try to force it to behave as a got client is probably out of scope. 

You can however do as you suggest and manually manage syncing and versioning of the sync files with git because they are basically plain text (unless you use e2ee). Having plugins is on the Joplin roadmap so in the future it will likely be possible to have specialized sync hooks.

**Comment by FilBot3** [comment]
There is a project that does Git syncing already for Android, GitJournal

* https://github.com/GitJournal/GitJournal

I've just been having issues with it not crashing when my BT keyboard disconnects from my tablet.

**Comment by QUIPQUOPRO** [comment]
Is this feature on the roadmap? Are there any other ideas for alternative storage providers?

**Comment by bitterbit** [comment]
Hi everyone,
I have been watching this issue for some time because I am always looking for tools to take notes that integrate with git. 

A few days ago I saw this forkable SQL database called Dolt.  https://github.com/dolthub/dolt#dolt. 
It is basically a SQL database that you can pull/push, branch, commit, diff, merge and etc... just like git. I didn't use it though so I don't know how stable it is, I had just saw this idea a few days ago and was blown away by the potential of git for non-textual data. 

Bottom Line: In case Joplin uses an SQL backend this could be very powerful.  Might even make it easy to program a diff for end-to-end encrypted notes. I hope someone finds this interesting or useful.
